packages/extension-driver-redshift/src/lib/redshiftDataSource.ts

```diff
@@ -0,0 +1,184 @@
+  DataResult,
+  ExecuteOptions,
+  InternalError,
+  RequestParameter,
+  VulcanExtensionId,
+} from '@vulcan-sql/core';
+import { Readable } from 'stream';
+import { buildSQL } from './sqlBuilder';
+import { mapFromRedShiftTypeId } from './typeMapper';
+import {
+  RedshiftDataClient,
+  RedshiftDataClientConfig,
+  ExecuteStatementCommand,
+  ExecuteStatementCommandInput,
+  ExecuteStatementCommandOutput,
+  DescribeStatementCommandInput,
+  DescribeStatementResponse,
+  DescribeStatementCommand,
+  GetStatementResultCommandInput,
+  GetStatementResultCommand,
+  SqlParameter,
+} from '@aws-sdk/client-redshift-data';
+
+export type RedshiftOptions = RedshiftDataClientConfig & Omit<ExecuteStatementCommandInput, "Sql" | "Parameters">;
+
+type RedShiftDataRow = {
+  [column: string]: any;
+}
+
+@VulcanExtensionId('redshift')
+export class RedShiftDataSource extends DataSource<any, RedshiftOptions> {
+  private logger = this.getLogger();
+  private redshiftClientMapping = new Map<
+    string,
+    {
+      redshiftClient: RedshiftDataClient;
+      options?: RedshiftOptions;
+    }
+  >();
+  public override async onActivate() {
+    const profiles = this.getProfiles().values();
+    for (const profile of profiles) {
+      this.logger.debug(
+        `Initializing profile: ${profile.name} using redshift driver`
+      );
+
+      const redshiftClient = new RedshiftDataClient(profile.connection!);
+      this.redshiftClientMapping.set(profile.name, {
+        redshiftClient: redshiftClient,
+        options: profile.connection,
+      });
+
+      await this.testConnection(profile.name);
+      this.logger.debug(`Profile ${profile.name} initialized`);
+    }
+  }
+
+  public async execute({
+    statement: sql,
+    bindParams,
+    profileName,
+    operations,
+  }: ExecuteOptions): Promise<DataResult> {
+    this.checkProfileExist(profileName);
+    const { redshiftClient, options } = this.redshiftClientMapping.get(profileName)!;
+
+    try {
+      const sqlParams: SqlParameter[] = [];
+      bindParams.forEach((value, key) => {
+        sqlParams.push({ name: key.replace(':', ''), value: String(value) });
+      });
+
+      const builtSQL = buildSQL(sql, operations);
+      let executeStatementCommandParams: ExecuteStatementCommandInput = {
+        Sql: builtSQL,
+        Database: options!.Database,
+        WorkgroupName: options!.WorkgroupName,
+      };
+      if (sqlParams.length) {
+        executeStatementCommandParams = {...executeStatementCommandParams, Parameters: sqlParams}
+      }
+
+      const executeStatementCommand = new ExecuteStatementCommand(executeStatementCommandParams);
+      const statementCommandResult = await redshiftClient.send(executeStatementCommand);
+      return await this.getResultFromExecuteStatement(statementCommandResult, redshiftClient);
+    } catch (e: any) {
+      this.logger.debug(
+        `Errors occurred, release connection from ${profileName}`
+      );
+      redshiftClient.destroy();
```

Comment:
Does `.destory` function here close the connection? We might reuse these clients so we should not close connections when error occurs.

Reply:
I've deleted `.destroy` function